awx_collection/plugins/module_utils/tower_api.py

```diff
@@ -460,20 +458,42 @@ def delete_if_needed(self, existing_item, handle_response=True, on_delete=None):
                 else:
                     self.fail_json(msg="Unable to delete {0} {1}: {2}".format(item_type, item_name, response['status_code']))
         else:
-            if not handle_response:
-                return None
+            self.exit_json(**self.json_output)
+
+    def modify_associations(self, association_endpoint, new_association_list):
+        # First get the existing associations
+        response = self.get_all_endpoint(association_endpoint)
+        existing_associated_ids = []
+        for exisitng_association in response['json']['results']:
+            existing_associated_ids.append( exisitng_association['id'] )
+
+        # Disassociate anything that is in existing_associated_ids but not in new_association_list
+        ids_to_remove = list(set(existing_associated_ids) - set(new_association_list))
+        for an_id in ids_to_remove:
+            response = self.post_endpoint(association_endpoint, **{'data': { 'id': int(an_id), 'disassociate': True }})
+            if response['status_code'] == 204:
+                self.json_output['changed'] = True
             else:
-                self.exit_json(**self.json_output)
+                self.fail_json(msg="Failed to disassociate item {0}".format(response['json']['detail']))
```

Comment:
what response codes cause us to fall into this exit conditions? 400s and the like are handled by the make_request method so shouldn't wind up here.

Do we even know that `detail` will be a key when this happens? And if we could show the response code that would also be ideal.

Reply:
If you pass into the API an invalid credential to associate or disassociate you get a 400 back:
```
HTTP 400 Bad Request
Allow: GET, POST, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept
X-API-Node: localhost
X-API-Time: 0.021s

{
    "detail": "Credential matching query does not exist."
}
```

In make_request, a 400 is not considered a fatal error and the response is returned from make_request back up the stack to allow the calling method to decide what to do with it.